url147ce07ebe_JoshuaKissoon_Kademlia_tgz-pJ8-kademlia_simulations_RoutingTableStateTestingJ8/src/kademlia/dht/DHT.java

```diff
@@ -0,0 +1,265 @@
+        {
+            KademliaStorageEntryMetadata current = this.contentManager.get(content.getContentMetadata());
+
+            /* update the last republished time */
+            current.updateLastRepublished();
+
+            if (current.getLastUpdatedTimestamp() >= content.getContentMetadata().getLastUpdatedTimestamp())
+            {
+                /* We have the current content, no need to update it! just leave this method now */
+                return false;
+            }
+            else
+            {
+                /* We have this content, but not the latest version, lets delete it so the new version will be added below */
+                try
+                {
+                    //System.out.println("Removing older content to update it");
+                    this.remove(content.getContentMetadata());
+                }
+                catch (ContentNotFoundException ex)
+                {
+                    /* This won't ever happen at this point since we only get here if the content is found, lets ignore it  */
+                }
+            }
+        }
+
+        /**
+         * If we got here means we don't have this content, or we need to update the content
+         * If we need to update the content, the code above would've already deleted it, so we just need to re-add it
+         */
+        try
+        {
+            //System.out.println("Adding new content.");
+            /* Keep track of this content in the entries manager */
+            KademliaStorageEntryMetadata sEntry = this.contentManager.put(content.getContentMetadata());
+
+            /* Now we store the content locally in a file */
+            String contentStorageFolder = this.getContentStorageFolderName(content.getContentMetadata().getKey());
+
+            try (FileOutputStream fout = new FileOutputStream(contentStorageFolder + File.separator + sEntry.hashCode() + ".kct");
+                    DataOutputStream dout = new DataOutputStream(fout))
+            {
+                this.getSerializer().write(content, dout);
+            }
+            return true;
+        }
+        catch (ContentExistException e)
+        {
+            /**
+             * Content already exist on the DHT
+             * This won't happen because above takes care of removing the content if it's older and needs to be updated,
+             * or returning if we already have the current content version.
+             */
+            return false;
+        }
+    }
+
+    @Override
+    public boolean store(KadContent content) throws IOException
+    {
+        return this.store(new JKademliaStorageEntry(content));
+    }
+
+    @Override
+    public JKademliaStorageEntry retrieve(KademliaId key, int hashCode) throws FileNotFoundException, IOException, ClassNotFoundException
+    {
+        String folder = this.getContentStorageFolderName(key);
+        DataInputStream din = new DataInputStream(new FileInputStream(folder + File.separator + hashCode + ".kct"));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the following resource: *din*. Currently, there are execution paths that do not contain closure statements, for example, when KadSerializer.read throws an exception. Ensure *din* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).